### DISCUSSION TITLE
How to change Edge attributes on hover/active?

### DISCUSSION THREAD

**Discussion by CSworenI** [xyflow/xyflow#4496] [Q&A] [answered]
So I understand that you can use things like `stroke` to change the color, `strokeWidth` to change the thickness, and `animated` to make the edge move, but how would someone go about these kinds of change on mouse `hover` or on click (`active`)?

**Comment by Santoshsohani** [comment]
https://github.com/user-attachments/assets/dc869565-5668-4c91-bf7f-86ba2c2b9380



- To change the color or the styles of edge, I have used `onEdgeMouseEnter` and `onEdgeMouseLeave`, As you can see the initial color for the edge is `red` and on hover the color changes to `green` with `animation` as well.
- Based on the above two events I have set the edges by changing their styles

```js
import React, { useState } from 'react';
import {
  ReactFlow,
  useNodesState,
  useEdgesState,
  MiniMap,
  Controls,
  Background,
} from '@xyflow/react';
import '@xyflow/react/dist/style.css';

const initialNodes = [
  { id: '1', data: { label: 'Node 1' }, position: { x: 200, y: 0 } },
  { id: '2', data: { label: 'Node 2' }, position: { x: 100, y: 100 } },
  { id: '3', data: { label: 'Node 3' }, position: { x: 300, y: 100 } },
];

const initialEdges = [
  { id: 'e1-2', source: '1', target: '2', style: { stroke: 'red' } },
  { id: 'e1-3', source: '1', target: '3', style: { stroke: 'red' } },
];

const App = () => {
  const [nodes, setNodes, onNodesChange] = useNodesState(initialNodes);
  const [edges, setEdges, onEdgesChange] = useEdgesState(initialEdges);
  const [currentEdge, setCurrentEdge] = useState(null);

  const handleEdgeChange = (e, edgeProp) => {
    setCurrentEdge(edgeProp.id);
    setEdges((edges) =>
      edges.map((edge) => {
        if (edge.id === edgeProp.id) {
          return { ...edge, style: { stroke: 'green' }, animated: true };
        } else {
          return edge;
        }
      })
    );
  };

  const handleEdgeMouseLeave = (e, edge) => {
    setEdges((edges) =>
      edges.map((edg) => {
        if (edg.id === edge.id) {
          return { ...edg, style: { stroke: 'red' }, animated: false };
        } else {
          return edg;
        }
      })
    );
    setCurrentEdge(null);
  };

  return (
    <div className="w-[100vw] h-[100vh]">
      <ReactFlow
        nodes={nodes}
        edges={edges}
        onNodesChange={onNodesChange}
        onEdgesChange={onEdgesChange}
        onEdgeMouseEnter={(e, edge) => handleEdgeChange(e, edge)}
        onEdgeMouseLeave={(e, edge) => handleEdgeMouseLeave(e, edge)}
      >
        <MiniMap />
        <Controls />
        <Background variant="dots" />
      </ReactFlow>
    </div>
  );
};

export default App;
```
- Check out the code

Hope this helps! 😎🚀

**Reply by bcakmakoglu** [reply, marked as answer]
Nice, although I think you can simplify this by using `updateEdge` instead of using `setEdges` :)

```tsx
const { updateEdge } = useReactFlow()

// ...

<ReactFlow
	  onEdgeMouseEnter={(_, edge) => updateEdge(edge.id, { style: { stroke: "green" } }) }
      onEdgeMouseLeave={(_, edge) => updateEdge(edge.id, { style: {} }) }
```

@CSworenI as for animating an edge when it's selected, I'd say the easiest way is to copy a little bit of the css of ReactFlow and just do this

```css
.react-flow__edge.selected path {
  stroke-dasharray: 5;
  animation: dashdraw 0.5s linear infinite;
}
```

**Reply by Santoshsohani** [reply]
Thank you so much! I'm working on some projects using XYFlow, got to learn about this approach :)

**Reply by CSworenI** [reply]
In trying the `updateEdge` approach, I'm getting the [Can't create edge](https://reactflow.dev/learn/troubleshooting#cant-create-edge-an-edge-needs-a-source-and-a-target) warning. My edge looks like this (console logged):
```
{
    "id": "e18-19", "source": "18", "target": "19", "type": "smoothstep", "sourceHandle": "out",
    "targetHandle": "bin", "label": "WIP", "markerEnd": { "type": "arrowclosed" }, "animated": true,
    "style": { "strokeWidth": 4, "stroke": "#fbbf24" }
}
```

I don't think the shape of it is the issue. Here's how I'm using `updateEdge`, maybe you can redirect me in how I might be doing it wrong?

```
const [oldEdgeStyle, setOldEdgeStyle] = useState(null);

const onEdgeHover = useCallback((edge) => {
	setOldEdgeStyle(edge.style);
	updateEdge(edge.id, { style: { stroke: 'lightgray', strokeWidth: 6, animated: true } });
}, []);

<ReactFlow
	onEdgeMouseEnter={(_, edge) => onEdgeHover(edge)}
        onEdgeMouseLeave={(_, edge) => updateEdge(edge.id, { style: {...oldEdgeStyle} })}
>
...
</ReactFlow>
```

**Reply by bcakmakoglu** [reply]
Are you on v12 and using `updateEdge` from `useReactFlow`?

In v11 there was also an `updateEdge` function but it's an export from `reactflow` and meant to *reconnect* an edge, thus in v12 this has been renamed to `reconnectEdge`

**Reply by CSworenI** [reply]
@bcakmakoglu Ok, so I updated to latest, changed from the old import to the `useReactFlow()`, and now I'm getting an error that `updateEdge` is not a function. I even integrated my change mechanism into the `onEdgeMouseEnter` prop of `ReactFlow` with the same issue. What would you suggest?

```
const [oldEdgeStyle, setOldEdgeStyle] = useState(null),
    { updateEdge } = useReactFlow();
... <ReactFlow
onEdgeMouseEnter={(_, edge) => {
    setOldEdgeStyle(edge.style);
    updateEdge(edge.id, { style: { stroke: 'lightgray', strokeWidth: 6, animated: true } });
} }
onEdgeMouseLeave={(_, edge) => updateEdge(edge.id, { style: {...oldEdgeStyle} })}
>...</ReactFlow>
```
![image](https://github.com/user-attachments/assets/1b5b4ef6-4396-4808-8355-29add4e87aab)

**Reply by bcakmakoglu** [reply]
Looks like you simply aren't on v12, make sure you actually installed v12 and you import from `@xyflow/react`.

**Reply by bcakmakoglu** [reply]
[Also read the migration guide for v12.](https://reactflow.dev/learn/troubleshooting/migrate-to-v12)

**Reply by CSworenI** [reply]
That explains it. Somehow I missed seeing that the library name changed for v12.

**Comment by corinaivn** [comment]
Nice example! I was wondering if you could make this work with active or selected state instead of hover. Thanks!

**Reply by CSworenI** [reply]
It shouldn't look too different. You alerted me to the fact that with the `onEdgeMouseEnter` property, I omitted `...edge,` in the opening curly braces, which is probably the culprit for some of the other weird behavior I was seeing. Continue with that destructuring to manipulate active/selected states as you wish (but keep in mind it may not persist custom styling that way, which was why I asked this question in the first place). In your case, you might use something like:

```
   onEdgeMouseEnter={(_ edge) => {
      setOldEdge(edge); // save your custom properties/state
      updateEdge(edge.id, { ...edge, active: true, selected: false } });
   }}
   onEdgeMouseLeave={(_, edge) => updateEdge(edge.id, oldEdge))} // reapply previous custom properties/state
```
